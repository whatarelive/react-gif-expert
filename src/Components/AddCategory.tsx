// Librerias de React.
import React, {useState} from "react";
// Types.
import {AddCategoryProps} from "../Types/types";


export const AddCategory = ( { onNewCategory }: AddCategoryProps ) => {

    // Hook useState : permite dal un estado al texto del input.
    const [ inputValue, setInputValue ] = useState( String );

    // Funcion para extraer el texto del input y guardarlo en el useState.
    const onInputChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => setInputValue( target.value );

    // Funcion que enviar el valor de la nueva categoria al componente padre.
    const onSubmit = (event: React.FormEvent) => {
        event.preventDefault();

        const text = inputValue.trim();
        
        if( text.length > 0 ) {
            onNewCategory( text );
            setInputValue('');
        }
    };

    // Renderizado del componente.
    return (
        <form onSubmit={ onSubmit }>
            <input
                type="text"
                placeholder="Buscar Gifs"
                value={ inputValue }
                onChange={ onInputChange }
            />
        </form>
    );
};