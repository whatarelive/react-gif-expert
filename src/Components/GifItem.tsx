import { GifType } from "../Types/types"

export const GifItem = ( image: GifType ) => {

    return (
         <div className="card-item">          
            <img src={ image.url } alt={ image.title } /> 
            <p>{ image.title }</p>
        </div>      
    )
}
