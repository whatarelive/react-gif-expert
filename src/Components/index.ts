// Archivo de barril para todas las importaciones de la carpeta componentes.
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';