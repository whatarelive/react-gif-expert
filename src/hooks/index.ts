export * from './useCategory';
export * from './useFecthGifs';