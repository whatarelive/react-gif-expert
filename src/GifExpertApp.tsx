// Librerias de React.
import {useState} from "react";
// Componentes hijos.
import { AddCategory, GifGrid } from "./Components";


export const GifExpertApp = () => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState(['One Punch']);

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if ( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories ]);
        }
    }

    // Funcion que crea el listado de elementos.
    const listItem = () => {
        return categories.map( ( category: string ) => {
           return <GifGrid key={ category } category={ category }/>;
        });
    }

    return (
        <>
            <h3>Gif Expert App</h3>
            <AddCategory onNewCategory={ onAddCategory }/>
            <ol>{ listItem() }</ol>
        </>
    );
};