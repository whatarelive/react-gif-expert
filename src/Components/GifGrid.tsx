// Custom Hook
import { useFecthGifs } from "../hooks/useFecthGifs";
// Componetes
import { GifItem } from "./index";
// Types.
import {GifGridProps} from "../Types/types";


export const GifGrid = ({ category }: GifGridProps ) => {

    const { images, isLoading } = useFecthGifs( category );
    
    // Renderizado del componente.
    return (
        <>
            <h3>{ category }</h3>

            {
                isLoading && ( <h2>Cargando.....</h2>)
            }

            <div className="card-grid"> 
            { 
                images.map( images => <GifItem key={ images.id } { ...images }/>)
            }
            </div> 
        </>
    );
};
