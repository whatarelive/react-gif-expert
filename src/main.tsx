// Librerias de React.
import React from 'react'
import ReactDOM from 'react-dom/client'
// Estilos CSS. 
import "./Styles/index.css";
// Componentes propios. 
import {GifExpertApp} from "./GifExpertApp.tsx";

const root = document.getElementById('root');

if (root instanceof HTMLDivElement) {

    ReactDOM.createRoot(root).render(
        <React.StrictMode>
                <GifExpertApp/>
        </React.StrictMode>
    );
} 