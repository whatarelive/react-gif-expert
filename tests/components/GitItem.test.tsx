import React from "react";
import { render, screen } from "@testing-library/react";
import { GifItem } from '../../src/Components';

describe('Pruebas en el componente <GifItem />', () => {

    // Elementos Random para probar.
    const title = 'Saitama';
    const url = 'https://one-punch.com/saitam.png';

    // Test #1: Sobre la UI del componente.
    test('debe de hacer match con el snapshot', () =>{
    
        // Se guarda el contenido del render del componente.
        const { container } = render( <GifItem  title={title} url={url} id={""} />);
        
        // Se crea una captura del elemento renderizado.
        expect( container ).toMatchSnapshot();
    });

    // Test #2: Sobre los datos a mostrar en la imagen del GifItem. 
    test('debe de mostrar la imagen con el URL y el ALT indicado', () => {

        render( <GifItem title={ title } url={ url } id="" />);
        
        // Nos permite ver el render del componente en la consola.
        //screen.debug();  

        // DesestructuraciÃ³n del HTMLImageElement.
        const { src, alt } = screen.getByRole('img') as HTMLImageElement;

        // Esperamos que el src sea igual al url.
        expect( src ).toBe( url );

        // Esperamos que el alt sea igual al title.
        expect( alt ).toBe( title );
    });

    test('debe mostrar el titulo en el en el componente ', () => {

        // Renderizamos el componente.
        render( <GifItem title={ title } url={ url } id="" />);

        // Esperamos que sea verdadero la aparicion del titulo como texto en el componente.
        expect( screen.getByText( title ) ).toBeTruthy();
    });
});