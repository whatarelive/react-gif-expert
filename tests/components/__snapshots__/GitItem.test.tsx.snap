// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifItem /> debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card-item"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitam.png"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
